File: rpi/llm/providers/openai_provider.py
from typing import AsyncIterator

from loguru import logger

from llm.base import BaseLLM


class OpenAIProvider(BaseLLM):
    """OpenAI ChatGPT provider with streaming support."""

    def __init__(self, api_key: str, model: str = "gpt-4o-mini"):
        self.api_key = api_key
        self.model = model
        self._client = None

    def _ensure_client(self):
        if self._client is None:
            from openai import AsyncOpenAI
            self._client = AsyncOpenAI(api_key=self.api_key)

    async def stream(self, messages: list[dict]) -> AsyncIterator[str]:
        """Stream tokens from OpenAI API."""
        if not self.api_key:
            logger.error("OpenAI API key not configured.")
            yield "I need to be connected to the internet for this. Ask your parent to check the settings!"
            return

        self._ensure_client()

        try:
            response = await self._client.chat.completions.create(
                model=self.model,
                messages=messages,
                max_tokens=150,
                temperature=0.4,
                stream=True,
            )

            async for chunk in response:
                delta = chunk.choices[0].delta if chunk.choices else None
                if delta and delta.content:
                    yield delta.content

        except Exception as e:
            logger.error("OpenAI streaming error: {}", e)
            yield "Hmm, I couldn't think of an answer right now. Let's try again!"

File: android_app/lib/core/llm/openai_provider.dart
import 'dart:async';
import 'dart:convert';

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';

import 'llm_provider.dart';

/// OpenAI GPT provider with streaming support.
class OpenAIProvider implements LlmProvider {
  final Dio _dio = Dio();
  String apiKey;
  final String model;

  OpenAIProvider({
    required this.apiKey,
    this.model = 'gpt-4o-mini',
  });

  @override
  String get name => 'openai';

  @override
  Stream<String> stream(List<Map<String, String>> messages) async* {
    if (apiKey.isEmpty) {
      debugPrint('[OpenAI] No API key');
      return;
    }

    try {
      debugPrint('[OpenAI] Streaming from $model...');
      final response = await _dio.post(
        'https://api.openai.com/v1/chat/completions',
        data: {
          'model': model,
          'messages': messages,
          'stream': true,
          'max_tokens': 300,
          'temperature': 0.7,
        },
        options: Options(
          headers: {
            'Authorization': 'Bearer $apiKey',
            'Content-Type': 'application/json',
          },
          responseType: ResponseType.stream,
          sendTimeout: const Duration(seconds: 30),
          receiveTimeout: const Duration(seconds: 60),
        ),
      );

      final stream = response.data.stream as Stream<List<int>>;
      String buffer = '';

      await for (final chunk in stream) {
        buffer += utf8.decode(chunk);

        // SSE format: each event is "data: {...}\n\n"
        while (buffer.contains('\n')) {
          final idx = buffer.indexOf('\n');
          final line = buffer.substring(0, idx).trim();
          buffer = buffer.substring(idx + 1);

          if (line.isEmpty) continue;
          if (line == 'data: [DONE]') return;
          if (!line.startsWith('data: ')) continue;

          try {
            final json = jsonDecode(line.substring(6));
            final delta = json['choices']?[0]?['delta']?['content'];
            if (delta != null && delta is String && delta.isNotEmpty) {
              yield delta;
            }
          } catch (_) {
            // Skip malformed JSON
          }
        }
      }
    } catch (e) {
      debugPrint('[OpenAI] Stream error: $e');
    }
  }
}
